### PR TITLE
NPM/CommonJS support

### LINKED ISSUE #51: Install via npm doesn't seem trustworthy
I'm planning on having a browserified build that includes this as a dependency. There's a `package.json` in the project root, so it seems like this would work via an NPM install. Looking at `package.json`, it looks like plugin has been registered on NPM with "jquery-plugin". 

npm installing that gives me the package, but I think there are a few issues here:

1) `jquery` and `addressfield.json` aren't listened as runtime dependencies, so I don't get them with the install (unlike if I install this with bower). node will not find these dependencies at runtime, so this will likely break on a browserified build that happens to not include jquery as a separate dependency.
2) in `bower.json`, I see the main file as `dist/jquery.addressfield.js`. I don't have this however:

```shell
example $ ls -lh node_modules/jquery-plugin/dist/
total 16
-rw-r--r--  1 valvarez  TSI\Domain Users   842B Sep  7  2012 foo.js
-rw-r--r--  1 valvarez  TSI\Domain Users   435B Sep  7  2012 foo.min.js
```

3) this is more of a suggestion, but could we make the package name on NPM a bit more specific, something like "jquery.addressfield" instead of "jquery-plugin"?
4) I get development files in the distribution:

```shell
example $ ls -lh node_modules/jquery-plugin/libs/
total 8
drwxr-xr-x  3 valvarez  TSI\Domain Users   102B Sep  7  2012 jquery
-rw-r--r--  1 valvarez  TSI\Domain Users   506B Sep  7  2012 jquery-loader.js
drwxr-xr-x  4 valvarez  TSI\Domain Users   136B Jan 17 19:33 qunit
example $ ls -lh node_modules/jquery-plugin/src/
total 8
-rw-r--r--  1 valvarez  TSI\Domain Users   851B Sep  7  2012 foo.js
example $ ls -lh node_modules/jquery-plugin/test/
total 16
-rw-r--r--  1 valvarez  TSI\Domain Users   1.1K Sep  7  2012 foo.html
-rw-r--r--  1 valvarez  TSI\Domain Users   1.7K Sep  7  2012 foo_test.js
```

I also get a `foo.jquery.json` in my project root. Do I need these on my disk?

Proposed solutions:

- exclude unneeded files in `.npmignore`
- rebuild and republish to npm
- change npm package name
- enumerate jquery and addressfield.json in dependencies or peerDependencies

### PR DESCRIPTION
**Pull request by yangmillstheory**
#### Summary

Fixes #51. Basically:

* distribute through NPM
* consolidate build (stuff removed from `libs` that are now pulled in at build time via NPM)
* support CommonJS environments

#### Details

* made `grunt-cli` a `devDependency` for people that don't want to use/share a global installation
* pull latest `addressfield.json` as a `dependency` from the git repo (should that be versioned?); move out of source control
* pull `jquery` as a `dependency` from NPM; move out of source control
* move test helpers into `test/` and out of `lib/`

I verified this works with browserify. Here's the test setup:

```shell
test $ ls -l node_modules/ | grep jquery.address
drwxr-xr-x   7 valvarez  62064481  238 Feb  6 08:57 jquery.addressfield

test $ ls -l
total 1504
-rw-r--r--    1 valvarez  62064481  755803 Feb  6 09:20 bundle.js
-rw-r--r--    1 valvarez  62064481     120 Feb  6 08:55 entry.js
-rw-r--r--    1 valvarez  62064481     134 Feb  5 22:03 index.html
drwxr-xr-x  124 root      62064481    4216 Feb  6 08:57 node_modules
-rw-r--r--    1 root      62064481     256 Feb  5 22:01 package.json

test $ more entry.js
var $ = require('jquery');
require('jquery.addressfield');


console.log(typeof $.fn.addressfield);
```

Bundle it up:
```shell
test $ node_modules/.bin/browserify entry.js --debug > bundle.js # create the bundle
```

Here's `index.html`:

```shell
test $ more index.html

<!doctype html>
<html>
<head>
  <title>jquery.addressfield</title>
</head>
<body>
  <script src="bundle.js"></script>
</body>
</html>
```
Open up `index.html` in your browser; you get `function`.

@iamEAP What version would you like to start with in `package.json`?

### CODE REVIEW

File: test/.jshintrc [hunk 1]
```diff
@@ -11,6 +11,7 @@
   "boss": true,
   "eqnull": true,
   "browser": true,
+  "evil": true,
```
Comment (yangmillstheory) [inline test/.jshintrc]: We need this so `document.write` lints since `jquery-loader` is now in `test/`.

File: src/jquery.addressfield.js [hunk 2]
```diff
@@ -4,9 +4,13 @@
...
-
-(function ($) {
-
+(function(factory) {
+  if (typeof module === "object" && typeof module.exports === "object") {
+    factory(require("jquery"));
+  } else {
+    factory(jQuery);
+  }
+}(function factory($) {
```
Comment (yangmillstheory) [inline src/jquery.addressfield.js]: For the UMD snippet, check [here](http://blog.npmjs.org/post/112712169830/making-your-jquery-plugin-work-better-with-npm).

File: package.json [hunk 3]
```diff
@@ -1,21 +1,42 @@
...
     "grunt-contrib-concat": "~0.3.0",
+    "grunt-contrib-connect": "~0.7.1",
     "grunt-contrib-copy": "~0.6.0",
+    "grunt-contrib-jshint": "~0.6.0",
     "grunt-contrib-uglify": "~0.2.0",
     "grunt-contrib-watch": "~0.4.0",
-    "grunt-contrib-clean": "~0.4.0",
-    "grunt": "~0.4.4",
-    "grunt-contrib-connect": "~0.7.1"
+    "grunt-qunit-istanbul": "~0.4.5"
```
Comment (yangmillstheory) [inline package.json]: My version of NPM re-ordered these automatically. Sorry!

File: package.json [hunk 4]
```diff
@@ -1,21 +1,42 @@
...
+  },
+  "dependencies": {
+    "jquery": ">=1.3.2",
+    "addressfield.json": "tableau-mkt/addressfield.json"
```
Comment (iamEAP) [inline package.json]: Yes, this should probably be versioned.  How about `tableau-mkt/addressfield.json#1.1`?

File: package.json (hunk 3 above)
Comment (iamEAP) [inline package.json]: Any reason not to dump the `libs` directory completely and just have a `devDependency` on `qunitjs ~1.11.0`?

**Comment by iamEAP** [comment]
@yangmillstheory Just a couple notes above.  Otherwise, how about we start this at `1.1.0-dev.0`?

File: package.json (hunk 4 above)
Comment (yangmillstheory) [inline package.json]: ad04855

File: package.json (hunk 3 above)
Comment (yangmillstheory) [inline package.json]: Ah, good point. This repository is getting a lot lighter...

23cdd1c

**Comment by yangmillstheory** [comment]
So one thing I noticed is that [there's a task](https://github.com/tableau-mkt/jquery.addressfield/blob/master/Gruntfile.js#L38) that copies `libs/addressfield.json/build/addressfield.min.json` into `dist/` at build time.

Should we keep this behavior for backwards-compatibility? I'd like to copy over the version from `node_modules`, but installing from git doesn't give me the minified version.

**Comment by iamEAP** [comment]
Hmm...  I think it's fine to break that functionality.  Shipping with the un-minified version seems good enough.

**Comment by yangmillstheory** [comment]
Ah ha! 

I found this in `.postinstall.sh`.

>#!/bin/bash
>echo "This is a temporary workaround until bower package publishing is a thing."
>echo "If you run into problems, just build addressfield.json."
>sleep 5

>#Build addressfield.json
>cd libs/addressfield.json
>npm install
>grunt

So this is what installs and builds `addressfield.min.json`. Later, grunt copies it over to `dist/`.

If we're not going to build & distribute `addressfield.min.json`, then I think we can get rid of `.postinstall.sh` as well? 

Also, is there any reason not to .gitignore `libs/` (basically `bower_components/`)?

**Comment by iamEAP** [comment]
I'm a little reluctant to completely get rid of the addressfield.json dependency; it's not a "requirement," strictly speaking, but it helps people get up and running quickly.  Seems annoying for every developer to discover that they need this other complex config file and then figure out how to load it their own way.

I'm completely fine with removing stuff that invokes build steps for addressfield.json though.  All that its build process does is validate syntax (which shouldn't be an issue if requiring from source or a package repo) and "minify" (insofar as JSON can be minified).

With that in mind, seems safe to remove:
- `.posinstall.sh`
- grunt task `copy:main`
- NPM dependency `grunt-contrib-copy`

But let's keep the dependency in both `package.json` and `bower.json`.

Once that's done, I believe there won't be anything in `libs` and we can probably just remove it entirely--no need to gitignore.

Seems fine to remove `devDependencies` from `bower.json`.

**Comment by yangmillstheory** [comment]
You got it. See 9feb6ac & 83df45a.

**Comment by iamEAP** [comment]
Thanks @yangmillstheory!  Super happy to get this in.  Diff looks good.  Will publish a 1.1.0 soon-ish to npm.

**Comment by iamEAP** [comment]
https://www.npmjs.com/package/jquery.addressfield

**Comment by yangmillstheory** [comment]
Nice! What kind of testing did you do before publishing?

**Comment by iamEAP** [comment]
I rely entirely on the automated tests.  May want to automate the test you outlined in the description.

**Pull request merged**